### PR TITLE
Sketch 'gcloud.pubsub' API.

### PR DESCRIPTION
**Pull request by tseaver**
Rather than starting from code, as #237 does, this PR starts with "science fiction" API docs:  I think it might be easier to discuss the intended API in this form.

### CODE REVIEW

**Comment by dhermes** [comment]
Do you have an HTML built version hosted somewhere?

**Comment by tseaver** [comment]
Nope.  I actually find the ReST easier to read.

**Comment by dhermes** [comment]
The link is not at a top-level:
screen_shot_086
(This may be because the subpackage doesn't exist.)

File: docs/pubsub-api.rst [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+Create a new topic for an explicit project:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.create_topic("topic_name", project_id='my.project')
+   >>> topic.name
+   'topic_name'
+
+Fetch an extant topic for the default project:
```
Comment (dhermes) [inline docs/pubsub-api.rst]: "extant" is a bit fancy for dev docs. What's wrong with "existing"?

File: docs/pubsub-api.rst [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+   >>> topic = pubsub.create_topic("topic_name")
+   >>> topic.name
+   'topic_name'
+
+Create a new topic for an explicit project:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.create_topic("topic_name", project_id='my.project')
```
Comment (dhermes) [inline docs/pubsub-api.rst]: Single and double quotes mixed. (I'm trying to keep the number of nit comments minimal.)

File: docs/pubsub-api.rst [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+   ...     topic.publish('this is the first message_payload')
+   ...     topic.publish('this is the second message_payload',
+   ...                   attr1='value1', attr2='value2')
+   [<message_id1>, <message_id2>]
+
+
+Manage subscriptions to topics
+------------------------------
+
+Create a new "pull" subscription for a topic:
```
Comment (dhermes) [inline docs/pubsub-api.rst]: "pull" doesn't need to be in quotes. If we are worried they don't know what pull is, we can provide a link to the pubsub docs.

File: docs/pubsub-api.rst [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+   ...                                          ack_deadline=90)
+
+Create a new "push" subscription for a topic:
+
+.. doctest::
+
+   >>> ENDPOINT = 'https://example.com/hook'
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.create_subscription('subscription_name',
```
Comment (dhermes) [inline docs/pubsub-api.rst]: If we provide this, do we want to provide tools to verify the URI? Or at least a link out?

File: docs/pubsub-api.rst [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.get_subscription('subscription_name')
+   >>> [message.id for message in subscription.pull(max_messages=2)]
+   [<message_id1>, <message_id2>]
+
+Fetch messages for a "pull" subscription without blocking (none pending):
```
Comment (dhermes) [inline docs/pubsub-api.rst]: What is the underlying behavior here? How does it compare to explicitly specifying a `max_messages` value?

**Comment by dhermes** [comment]
Overall notes:

- I see no mentions of the relevant nouns as classes. Is that because those will end up in those module docs once in existence?
- This seems a bit verbose for a package landing page (but fine for just mocking out functionality)

File: docs/pubsub-api.rst (hunk 1 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: The connotation of [extant](http://www.merriam-webster.com/dictionary/extant) is that a thing (document, for instance), which might have been destroyed, still exists.  Given that topics might be destroyed between the time they were initially created and the time the `get_topic()` call is made, that connotation is precisely what I meant.

File: docs/pubsub-api.rst (hunk 4 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: Do you mean link to the PubSub docs on how to set up an "allowed" domain?

File: docs/pubsub-api.rst (hunk 1 above)
Comment (dhermes) [inline docs/pubsub-api.rst]: I'm not saying the word isn't correct or what you meant, I'm saying it's a word people don't commonly see.

The connotation was lost on me, nuance right over my head. A simpler word choice (or more words) would appeal to more readers and make the concept clearer.

File: docs/pubsub-api.rst (hunk 5 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: Without `return_immediately`, the API call blocks if there are no messages available.  Without `max_messages`, you get the API-specified default for `max_messages`.

File: docs/pubsub-api.rst (hunk 2 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: 883574e

File: docs/pubsub-api.rst (hunk 3 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: 9d82cf5

**Comment by tseaver** [comment]
>   I see no mentions of the relevant nouns as classes. Is that because those will end up in those module docs once in existence?

Presumably -- this document just tries to lay out the "normal path" for using the API, rather than the docstrings.

>  This seems a bit verbose for a package landing page (but fine for just mocking out functionality)

Sure.  Mostly, I'd like agreement that an API which looks like this would be sufficient to start implementing.  @jgeewax, is there somebody on the back-end team who should review?

File: docs/pubsub-api.rst (hunk 4 above)
Comment (dhermes) [inline docs/pubsub-api.rst]: Yeah: https://cloud.google.com/pubsub/prereqs#register

I re-read it and there is no way we could provide any helper code for verification. Browser required for 100% of it.

**Comment by dhermes** [comment]
@tmatsuo Would be a good start for someone very familiar with the PubSub API.

@tseaver Sorry I wasn't specific enough. My question was more geared toward asking if / why the "normal path" didn't ever use classes / the underlying nouns. (It could just be that the API is so simple we don't need the nouns. Verbs all day!)

**Comment by tseaver** [comment]
In the "main path" docs as I sketched them, I didn't see a need to make the user think about the `Topic` and `Subscription` classes at all:

- I don't expect any normal users ever to instantiate one directly (they have so little interesting state)
- The handful of methods are covered via their usage.

File: docs/pubsub-api.rst [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+Manage topics for a project
+---------------------------
+
+Create a new topic for the default project:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.create_topic("topic_name")
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Just a nit -- thought for most of the project we used single quotes (`'`). Not sure if that's true anymore...

File: docs/pubsub-api.rst [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+Attempt to fetch a non-extant topic:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> pubsub.get_topic('nonesuch')
+   Traceback (most recent call last):
+     File "<stdin>", line 1, in ?
+   NotFound: ...
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Should we raise an Error or return `None`? Just want to make sure this decision is on purpose.

File: docs/pubsub-api.rst [hunk 8]
```diff
@@ -0,0 +1,255 @@
+``gcloud.pubsub`` API
+=====================
+
+Connection / Authorization
+--------------------------
+
+- Inferred defaults used to create connection if none configured explicitly:
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Do we have an example of explicitly specifying credentials?

What about connecting to two different projects with two different sets of credentials?

Something like... 

```python
from gcloud import pubsub
project1 = pubsub.Connection(project_id='...', credentials_json='/path/to/key1.json')
topic = project1.create_topic('...')

project2 = pubsub.Connection(project_id='...', credentials_json='/path/to/key2.json')
topic = project2.create_topic('...')
```

..?

File: docs/pubsub-api.rst [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+   >>> [topic.name for topic in pubsub.list_topics(project_id="my.project")]
+   ['topic_name']
+
+Delete a topic:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> topic.delete()
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Should we offer a `pubsub.delete_topic('topic_name')` ? or just say to do `pubsub.get_topic('topic_name').delete()`

This would mean 2x the requests if you were deleting a lot of topics... Not sure how you guys feel about that.

File: docs/pubsub-api.rst [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+---------------------------
+
+Publish a single message to a topic, without attributes:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> topic.publish('this is the message_payload')
+   <message_id>
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Is there any use to having a `Message` object ? I'm guessing no?

File: docs/pubsub-api.rst [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+Manage subscriptions to topics
+------------------------------
+
+Create a new pull subscription for a topic:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
```
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Does the following call
pubsub.get_topic('topic_name')
issue a Cloud Pub/Sub API call?

It's unnecessary for just creating a new subscription.

File: docs/pubsub-api.rst [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+   ...               attr1='value1', attr2='value2')
+   <message_id>
+
+Publish a set of messages to a topic (as a single request):
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> with topic:
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: This is awesome. Huge +1.

File: docs/pubsub-api.rst [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+   ...     topic.publish('this is the first message_payload')
+   ...     topic.publish('this is the second message_payload',
+   ...                   attr1='value1', attr2='value2')
+   [<message_id1>, <message_id2>]
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Will these be guaranteed to be the same order?

File: docs/pubsub-api.rst [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+Manage subscriptions to topics
+------------------------------
+
+Create a new pull subscription for a topic:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.create_subscription('subscription_name')
```
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Does this library support the following case?

I have 'topic-a' in 'project-a', I want to create 'sub-b' in 'project-b', attached to the 'topic-a' in 'project-a'.

File: docs/pubsub-api.rst (hunk 4 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: Yikes that sucks.

At the very least, when we hit the error of "invalid domain" we should point people to that link... And our docs should have a note explaining the story, yea?

File: docs/pubsub-api.rst [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+Attempt to get a non-extant subscription for a topic:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.get_subscription('nonesuch')
+   Traceback (most recent call last):
+     File "<stdin>", line 1, in ?
+   NotFound: ...
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Exception or `None`?

File: docs/pubsub-api.rst [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+
+Update the ACK deadline for a subscription:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.get_subscription('subscription_name')
+   >>> subscription.modify_ack_deadline(90)
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Just out of curiosity, would it be more clear if we just let you "update" the arguments you passed into the constructor?

ie:

```python
subscription = topic.create_subscription('subscription_name', ack_deadline=90)
subscription.update(ack_deadline=120, push_endpoint=ENDPOINT)
subscription.update(push_endpoint=None)
```

File: docs/pubsub-api.rst [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.get_subscription('subscription_name')
+   >>> [message.id for message in subscription.pull()]
+   [<message_id1>, <message_id2>, ...]
+
+Fetch a limited number of pending messages for a pull subscription:
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Is there a way to provide tags here?

@tmatsuo : I thought we could tag messages and then pull messages grouped by tag.

File: docs/pubsub-api.rst [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+   [<message_id1>, <message_id2>]
+
+Fetch messages for a pull subscription without blocking (none pending):
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> topic = pubsub.get_topic('topic_name')
+   >>> subscription = topic.get_subscription('subscription_name')
+   >>> [message.id for message in subscription.pull(return_immediately=True)]
```
Comment (jgeewax) [inline docs/pubsub-api.rst]: Should we make this the default behavior and then provide a `wait=True` parameter?

```python
[message.id for message in sub.pull(wait=True)]
# Wait until there is a message
[message.id for message in sub.pull()]
[]
```

**Comment by jgeewax** [comment]
Sorry for the boatloads of comments. This looks really awesome though.

File: docs/pubsub-api.rst (hunk 16 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: modifyAckDeadline is for each message (associated with the given ackID). It's not modifying a global setting for a subscription.

File: docs/pubsub-api.rst (hunk 17 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: You can add 'attributes' (map[string]string) on messages.
Currently we don't provide filtering.

File: docs/pubsub-api.rst (hunk 14 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: @tmatsuo : Can you show us the code you'd want to write to accomplish that task?

File: docs/pubsub-api.rst [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+   >>> topic = pubsub.get_topic('topic_name', project_id="my.project")
+   >>> topic.name
+   'topic_name'
+
+Attempt to fetch a non-extant topic:
+
+.. doctest::
+
+   >>> from gcloud import pubsub
+   >>> pubsub.get_topic('nonesuch')
```
Comment (tmatsuo) [inline docs/pubsub-api.rst]: I think it's problematic if you issue a Cloud Pub/Sub API call when you do:
```python
pubsub.get_topic('nonesuch')
```
because it forces users to issue 2 requests for publishing/pulling or whatever. The first 'get' request is totally unnecessary.

Comment (tmatsuo) [inline docs/pubsub-api.rst]: and same for get_subscription()

File: docs/pubsub-api.rst (hunk 16 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: Wait, so you're saying it's not possible to create a subscription with no `ack_deadline` and then update the `ack_deadline` ?

[The docs](https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/subscriptions/modifyAckDeadline) are *super* misleading about this.

If I create a `subscription` that takes an `ackDeadlineSeconds` property and an API method called `modifyAckDeadline` on the `subscription` object -- I expect the `modifyAckDeadline` to ... change the one I provided.

In short, I'd expect...

- `POST projects/my-project/subscriptions/my-subscription:modifyAckDeadline` to change the ack deadline on the **subscription**.

- `POST projects/my-project/messages/message-ack-id-here:modifyAckDeadline` to change the ack deadline on the **message**.

The docs say that this message takes an `ackId` parameter, meaning it does as you say and modifies the ack deadline for JUST a single message. Was this an oversight?

File: docs/pubsub-api.rst (hunk 17 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: [The docs](https://cloud.google.com/pubsub/reference/rest/google/pubsub/v1beta2/PubsubMessage) say

"attributes" "map (key: string, value: string)" "Optional attributes for this message. This can be used to filter messages on the subscription."

File: docs/pubsub-api.rst (hunk 11 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: What if the topic doesn't exist? Shouldn't we check that it exists before letting you create a subscription on it?

File: docs/pubsub-api.rst (hunk 13 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Yes they should be. Otherwise it's useless IMO.

File: docs/pubsub-api.rst (hunk 11 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: In that case, the subsequent create_subscription() call will fail, and it seems fine to me.

File: docs/pubsub-api.rst (hunk 14 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: It can be:
```
from gcloud import pubsub
topic = pubsub.get_topic('topic-a', project_id='project-a')
subscription = topic.create_subscription('sub-b', project_id='project-b')
```

File: docs/pubsub-api.rst (hunk 4 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Agreed. Feel free to give feedback on the doc!

File: docs/pubsub-api.rst (hunk 19 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: Don't we want to check first if the topic exists? Or would you suggest something else?

File: docs/pubsub-api.rst (hunk 14 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: How do we handle authentication for something like that ?

File: docs/pubsub-api.rst (hunk 11 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: Wouldn't we want the failure to be as close to the issue as possible? Are you saying we should offer a `create_subscription` method directly that takes a topic or topic name?

```python
pubsub.create_subscription('subscription_name', topic_name='topic_name')
```

Which can throw a "Topic Not found" error?

File: docs/pubsub-api.rst (hunk 13 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: OK [double checked the docs](https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/topics/publish). We're good "The server-assigned ID of each published message, in the same order as the messages in the request. IDs are guaranteed to be unique within the topic."

File: docs/pubsub-api.rst (hunk 16 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: I understand that what you suggesting is beautiful in ideal world. We don't currently provide a way of modifying the global ackDeadlineSeconds on a subscription, but you can create a new one.

File: docs/pubsub-api.rst (hunk 17 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Well, we should remove the second sentence.

File: docs/pubsub-api.rst (hunk 16 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: I think we need to change this API to make more sense then. It seems to be extremely misleading. Who should we talk to about that?

File: docs/pubsub-api.rst (hunk 17 above)
Comment (jgeewax) [inline docs/pubsub-api.rst]: OK - How do we do that? Can you file a bug?

File: docs/pubsub-api.rst (hunk 19 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Please just suppressing any unnecessary API calls.
To publish, you don't need to call topics().get() first.
To pull, you don't need to call subscriptions().get() first.
etc, etc.

File: docs/pubsub-api.rst (hunk 14 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: Currently, if the caller has an 'editor' or greater permission of 'project-b', it just works.

File: docs/pubsub-api.rst (hunk 16 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: I'll e-mail with the internal group address, but I don't think dynamically changing the subscription level ackDeadlineSeconds is more important than modifying ackDeadlineSeconds on a single message.

File: docs/pubsub-api.rst (hunk 17 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: yep, I filed a bug

File: docs/pubsub-api.rst (hunk 19 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: AFAICT, except for the "existence check", there is really no reason for `topic.get()` to exist (the resource has no state beyond its name, which is embedded in the URL).

If we don't mean to be checking for the existence of the topic, then we could just replace `pubsub.get_topic()` with instantiation of `Topic`, e.g.:

```python
from gcloud.pubsub.topic import Topic
topic = Topic('topic_name')
```

File: docs/pubsub-api.rst (hunk 7 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: I meant to raise the exception (it matches the API semantics).

File: docs/pubsub-api.rst (hunk 9 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: If we move toward instantiating `Topic` w/o the `get_topic()` call, that concern should evaporate.

File: docs/pubsub-api.rst (hunk 10 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: I don't think so, at least not for `publish()` (there might be a reason for it in `pull()`, or in handling a `push`).

File: docs/pubsub-api.rst (hunk 11 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: If we instantiate `Topic` w/o `get_topic()`, then we can use its `create_subscription()` to get an instance of `Subscription`.   Or maybe we should instantiate `Subscription`, passing the `Topic`?  E.g.:

```python
from gcloud.pubsub.subscription import Subscription
from gcloud.pubsub.topic import Topic

topic = Topic('topic_name')
subscription = Subscription('subscription_name', topic)
```

File: docs/pubsub-api.rst (hunk 14 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: In the API, how does one create a subscription to a topic in a separate project?  In the [API docs for `subscription.create`](https://cloud.google.com/pubsub/reference/rest/v1beta2/projects/subscriptions/create), the URL for the PUT request contains the project ID of the subscription, but the payload contains only a topic name (no project ID).

File: docs/pubsub-api.rst (hunk 15 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: I meant it to raise an exception.  We may need to move to instantiating the `Subscription` object directly (to avoid the extra API requests, per @tmatsuo).

File: docs/pubsub-api.rst (hunk 19 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: If there's a way of just publishing without calling topics().get() it's fine for me.

File: docs/pubsub-api.rst (hunk 11 above)
Comment (tmatsuo) [inline docs/pubsub-api.rst]: LGTM

File: docs/pubsub-api.rst (hunk 16 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: I don't even understand the use case for modifying the deadline for a single message.  I would expect that if my app has pulled a message already, there won't be any way to know that I would miss the deadline in time to update it (I could just ACK it directly).

File: docs/pubsub-api.rst (hunk 18 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: We could do that, if that is the consensus.  I'd rather match the APIs semantics, myself.

File: docs/pubsub-api.rst (hunk 8 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: Because the story is going to be pretty much identical to the `datastore` / `storage` cases, I didn't include it in this sketch.

File: docs/pubsub-api.rst (hunk 6 above)
Comment (tseaver) [inline docs/pubsub-api.rst]: My own convention (which I fail to follow here) is to use double quotes for human-centric text (e.g., error messages), and single quotes for code-centric text (e.g., identifiers).

**Comment by tseaver** [comment]
#742 picks this up and implements the topic-related bits.
